Embed QR code scanner for syncing devices on onboarding

igor-sirotin opened the pull request:
FIxes https://github.com/status-im/status-desktop/issues/9865

Requires:
- https://github.com/status-im/status-desktop/pull/10088
- https://github.com/status-im/status-go/pull/3358


# What does the PR do

1. Embed `StatusQrCodeScanner` for "Setup with sync code" onboarding
1. Link `qzxing` as shared library
`qzxing` is built with many exceptions and `try/catch` blocks.
Currently catching doesn't work in the desktop app (though it works in sandbox).
When the library is build as shared one, it works. 
https://github.com/status-im/status-desktop/issues/9863 should fix this, for example.
1. Log error on `inputConnectionStringForBootstrappingFinished`
This should not happen, so I didn't implement any UI here
1. Hide user profile image until it's received from other device.

## Scanning different QR codes

https://user-images.githubusercontent.com/25482501/226880717-eb09126e-0754-4432-a765-bd17f610b93e.mov

## Scanning real QR code from mobile

https://user-images.githubusercontent.com/25482501/226881198-598b88b6-ca1c-469c-8d4d-b4164b63b636.mov

# `StatusQrCodeScanner`:
1. Added combo box for selecting a camera
3. Added "almost async" loading
Note [this comment in the code](https://github.com/status-im/status-desktop/blob/7eff03ec30670c799cf07d25ae006ad7acfb1f9c/ui/StatusQ/src/StatusQ/Components/StatusQrCodeScanner.qml#L15-L36) for more info.
4. Improved sandbox page

<img width="1325" alt="image" src="https://user-images.githubusercontent.com/25482501/226879596-f1c3731d-efe6-476e-b975-fd3744fab071.png">

# Important notes:

1. Asking for camera access on MacOS.
There're no such default handles in QML do detect if camera access was already gained.
This should be implemented in C++. I decided to postpone this feature and always show `Enable camera access` button.
This gratefully simplifies work for me and doesn't make much harm on UX.
Will implement it later.
1. There were some UI freezes in MacOS 13. 
Those are fixed in 13.1, as noted [here](https://bugreports.qt.io/browse/QTBUG-108530). I updated and tested with 13.2.
1. There's a [known Qt bug](https://bugreports.qt.io/browse/QTBUG-108018) on using [MacOS Continuity Camera feature](https://support.apple.com/guide/mac-help/use-iphone-as-a-webcam-mchl77879b8a/mac).
Currently we can't really do anything about it. The app will crash when choosing the iPhone camera in the combobox.

# Affected areas

Onboarding with sync code

igor-sirotin commented:
@glitchminer Here're some updates on syncing fixes 🙂 
I only tested this on MacOS, but this should definitely be tested on other systems.
Can I please ask you to check it on Windows? And Linux, if you have any 😄

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_multiplication_x: | 7eff03ec | [#1](https://ci.status.im/job/status-desktop/job/branches/job/tests-nim/job/PR-9981/1/) | 2023-03-22 11:00:20 | ~5 min | `tests-nim` | [:page_facing_up:`log`](https://ci.status.im/job/status-desktop/job/branches/job/tests-nim/job/PR-9981/1//consoleText) |

caybro commented on ui/StatusQ/sandbox/QrCodeScannerPage.qml:
```diff
@@ -0,0 +1,157 @@
+
+        ValueIndicator {
+            Layout.fillWidth: true
+            title: qsTr("Last tag:")
```
No need to translate stuff in sandbox (same below)

igor-sirotin replied:
Sure, removed 👍

caybro commented on ui/StatusQ/src/StatusQ/Components/StatusQrCodeScanner.qml:
```diff
@@ -81,91 +152,146 @@ Item {
+        Loader {
+            active: root.highlightContentZone
+            sourceComponent: Rectangle {
+                color: "blue"
```
"blue", not some color from `Theme.palette`?

igor-sirotin replied:
Well this one is just highlighting for testing purposes. Which is not used in the app actually.
I don't think it's needed to use any theme colors here 🙂

caybro commented on ui/StatusQ/src/StatusQ/Components/StatusQrCodeScanner.qml:
```diff
@@ -81,91 +152,146 @@ Item {
+                color: "hotpink"
+                opacity: 0.2
+                border.width: 3
+                border.color: "hotpink"
```
same here, "hotpink"

caybro reviewed (approved):
Looks good, some minor stuff only from my side

osmaczko commented on ui/StatusQ/src/StatusQ/Components/StatusQrCodeScanner.qml:
```diff
@@ -7,42 +7,112 @@ import QtGraphicalEffects 1.0
+    Therefore, we can't simply use Loader with `asyncronous` flag.
+    Neiter we can set `loading: loader.status === Loader.Loading` to this button.
+
+    To achieve desired points, I manually set `loading` property of the button
```
This can be mitigated by using Animator: https://github.com/status-im/status-desktop/pull/10012

glitchminer commented:
Hi @igor-sirotin , I had 1 crash on the app displaying the QR code when I attempted to sync a second time. Here is the console:

'''
INF 2023-03-29 14:45:26.859+01:00 history request completed                  topics="mailservers-service" tid=82587942 file=service.nim:183
Warning: qrc:/app/AppLayouts/Profile/popups/SetupSyncingPopup.qml:191:18: QML QQuickItem: Binding loop detected for property "implicitHeight" (qrc:/app/AppLayouts/Profile/popups/SetupSyncingPopup.qml:191, )
Warning: qrc:/app/AppLayouts/Profile/popups/SetupSyncingPopup.qml:191:18: QML QQuickItem: Binding loop detected for property "implicitHeight" (qrc:/app/AppLayouts/Profile/popups/SetupSyncingPopup.qml:191, )
Warning: qrc:/app/AppLayouts/Profile/popups/SetupSyncingPopup.qml:191:18: QML QQuickItem: Binding loop detected for property "implicitHeight" (qrc:/app/AppLayouts/Profile/popups/SetupSyncingPopup.qml:191, )
INF 2023-03-29 14:48:00.069+01:00 shutting down...                           topics="status-app" tid=82587942 file=nim_status_client.nim:161
INF 2023-03-29 14:48:00.070+01:00 logging out...                             topics="app-controller" tid=82587942 file=app_controller.nim:278
Warning: qrc:/app/mainui/Popups.qml:54: TypeError: Cannot call method 'indexOf' of undefined (qrc:/app/mainui/Popups.qml:54, )
Traceback (most recent call last)
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9981/src/nim_status_client.nim(197) nim_status_client
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9981/src/nim_status_client.nim(171) mainProc
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9981/src/app/core/main.nim(29) delete
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9981/src/app/core/signals/signals_manager.nim(22) delete
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9981/vendor/nimqml/src/nimqml/private/qobject.nim(64) delete
SIGBUS: Illegal storage access. (Attempt to read from nil?)
'''

glitchminer commented:
@igor-sirotin , here are the error states mentioned on the call. They should be in a separate issue right?

Not a sync code: Shows a different error to Figma https://www.figma.com/file/17fc13UBFvInrLgNUKJJg5/Kuba%E2%8E%9CDesktop?node-id=25869-519999&t=NLevci35MYWqvDOx-4
In app:
![image](https://user-images.githubusercontent.com/50769329/228587152-2ca12d70-afc2-4603-a121-0b1d5842e436.png)

---
Expired code: Error is not per Figma
![image](https://user-images.githubusercontent.com/50769329/228585129-e55ecf0c-1358-4b23-b05b-8ddd91387c03.png)

---
Code already used:  Not per Figma
![image](https://user-images.githubusercontent.com/50769329/228586579-ae37f22d-be2a-4e78-9da4-52041a4e2698.png)

---
Also noted that the device name was not available on confirmation:
![image](https://user-images.githubusercontent.com/50769329/228585566-019c6c19-6c3e-409d-bde6-892099791ef6.png)

caybro commented:
@igor-sirotin the timestamp (and date shown) is wrong if you never synced before:
![image](https://user-images.githubusercontent.com/5377645/228588876-5be32db2-9480-4a99-825f-bd69eefd1461.png)

(shows 7th December 1969; passing the timestamp as an `int` somewhere? ;)

glitchminer commented:
Got a mobile version installed now (StatusIm-Mobile-230329-030800-006b11-nightly-universal.apk).
While Desktop to Desktop QR appears to be recognised, Desktop > Mobile and Mobile > Desktop QR codes are not being picked up.

I also upgraded the MacOS version. There is still a slight freeze of a couple of seconds (sometimes beachball) when enabling the camera for the first time. This appears to be due to having multiple cameras attached, but it is significantly reduced from under the previous version.

igor-sirotin commented:
@glitchminer, thanks for the review! I've updated the PR.

## Crash on second sync

Could you please give more details on this one?
Is this after syncing an already synced account with Mobile->Desktop or Desktop->Mobile?
Or is it after scanning a QR code second time, after another sync code has already been scanned with error?

## "Not a QR code" error

I made this for consistency with text input of sync code.
We came to a consensus with design team. Figma and code updated.

## Expired code / Code already used

Indeed, I haven't implemented those messages yet.
Will open a separate issue for this, doesn't seem to be urgent.

## Device name was not available on confirmation

I'm tracking this here: https://github.com/status-im/status-desktop/issues/9930

## Sync with mobile

Updated to newer version of `status-go` and tested. Should work fine now 👌

igor-sirotin commented:
This PR now requires https://github.com/status-im/status-go/pull/3358 to be merged. 

It also requires this change: https://github.com/status-im/status-desktop/pull/10088
It's actually already in this PR, but I'll remove the duplication in 0ca518b07781b3a4d9e7ad7d426c8a288d60353b when https://github.com/status-im/status-desktop/pull/10088 is merged.

osmaczko commented:
I can confirm that the scanning feature works fine on Linux :+1: 

https://user-images.githubusercontent.com/33099791/228846432-1319d084-ea34-4c98-91f9-4dd1da4cd021.mp4

igor-sirotin commented on Makefile:
```diff
@@ -676,11 +684,11 @@ run-macos: nim_status_client
 
 tests-nim-linux: | $(DOTHERSIDE)
-	LD_LIBRARY_PATH="$(QT5_LIBDIR)" \
+	LD_LIBRARY_PATH="$(QT5_LIBDIR)":"$(QZXING_LIBDIR)" \
```
Yeap. `tests-nim` weren't passing without.

glitchminer reviewed (approved):
Tested, approved:

Successfully logged in using syncs via QR:
Desktop > Desktop
Desktop > Mobile
Mobile > Desktop

Update to "not a sync QR" message OK.

I think all other QA items raised on here have been addressed. The following have now been been moved:
Crash on second login to https://github.com/status-im/status-desktop/issues/10096
Failure to set up a new sync during first run after a sync to https://github.com/status-im/status-desktop/issues/10097

Noted some issues with sync itself to be tested further - eg. profile pic not showing until after restart.
